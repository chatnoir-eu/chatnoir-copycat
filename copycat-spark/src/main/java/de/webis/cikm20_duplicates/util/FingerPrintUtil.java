package de.webis.cikm20_duplicates.util;

import java.io.Serializable;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import de.aitools.ir.fingerprinting.hashfunction.SimHash;
import de.aitools.ir.fingerprinting.representation.HashVector;
import de.aitools.ir.fingerprinting.representation.HashVectorSha3;
import de.aitools.ir.fingerprinting.representer.Hash;
import de.webis.cikm20_duplicates.spark.eval.SparkEvaluateSimHashFeatures;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import de.webis.trec_ndd.util.NGramms;
import info.debatty.java.lsh.MinHash;
import lombok.experimental.UtilityClass;

/**
 * 
 * @author Maik Fr√∂be
 *
 */
@UtilityClass
public class FingerPrintUtil {
	public static interface Fingerprinter<T extends Comparable<T>> extends Serializable {
		public List<T> fingerprint(CollectionDocument doc);
		public List<T> fingerprint(List<String> features);
		
		public double similarity(List<T> a, List<T> b);
		public String fingerprinterName();
		
		public default double similarity(CollectionDocument a, CollectionDocument b) {
			List<T> fingerprintA = fingerprint(a);
			List<T> fingerprintB = fingerprint(b);
			
			return similarity(fingerprintA, fingerprintB);
		}
	}

	@SuppressWarnings("serial")
	public static Fingerprinter<Integer> minHashFingerPrinting(long seed) {
		int dict_size = 1;
		int size = 12;
		MinHash minHash = new MinHash(size, dict_size, seed);
		
		return new Fingerprinter<Integer>() {
			@Override
			public List<Integer> fingerprint(CollectionDocument doc) {
				return IntStream.of(minHash.signature(docToElementSet(doc))).boxed().collect(Collectors.toList());
			}
			
			@Override
			public List<Integer> fingerprint(List<String> features) {
				throw new RuntimeException();
			}

			@Override
			public double similarity(List<Integer> a, List<Integer> b) {
				int[] aArray = a.stream().mapToInt(i -> i).toArray();
				int[] bArray = b.stream().mapToInt(i -> i).toArray();
				
				return minHash.similarity(aArray, bArray);
			}
			
			public String fingerprinterName() {
				return "MinHashWithJavaHash";
			}
		};
	}
	
	@SuppressWarnings("serial")
	public static Fingerprinter<Integer> simHashFingerPrinting(int bitsInSimHash, int k) {
		return new Fingerprinter<Integer>() {

			@Override
			public List<Integer> fingerprint(CollectionDocument doc) {
				byte[] hash = hashVector(doc, bitsInSimHash);
				
				return HashTransformationUtil.hashToIntegers(hash, k);
			}

			@Override
			public List<Integer> fingerprint(List<String> features) {
				byte[] hash = hashVector(features, bitsInSimHash);
				
				return HashTransformationUtil.hashToIntegers(hash, k);
			}
			
			@Override
			public double similarity(List<Integer> a, List<Integer> b) {
				byte[] aArr = HashTransformationUtil.integersToHash(a);
				byte[] bArr = HashTransformationUtil.integersToHash(b);
				
				return 1.0 - (((double)Hash.getHammingDistance(aArr, bArr))/((double)aArr.length*8));
			}
			
			public String fingerprinterName() {
				return "64BitK3SimHashOneGramms";
			}
		};
	}
	
	@SuppressWarnings("serial")
	public static Fingerprinter<Integer> productionFingerpringint(int bitsInSimHash, int k) {
		return new Fingerprinter<Integer>() {

			@Override
			public List<Integer> fingerprint(CollectionDocument doc) {
				byte[] hash = hashVector(SparkEvaluateSimHashFeatures.theeAndFiveGramms(doc), bitsInSimHash);
				
				return HashTransformationUtil.hashToIntegers(hash, k);
			}

			@Override
			public List<Integer> fingerprint(List<String> features) {
				byte[] hash = hashVector(features, bitsInSimHash);
				
				return HashTransformationUtil.hashToIntegers(hash, k);
			}
			
			@Override
			public double similarity(List<Integer> a, List<Integer> b) {
				byte[] aArr = HashTransformationUtil.integersToHash(a);
				byte[] bArr = HashTransformationUtil.integersToHash(b);
				
				return 1.0 - (((double)Hash.getHammingDistance(aArr, bArr))/((double)aArr.length*8));
			}
			
			public String fingerprinterName() {
				return "64BitK3SimHashThreeAndFiveGramms";
			}
		};
	}
	
	static byte[] hashVector(CollectionDocument doc, int bitsInSimHash) {
		return new SimHash().hash(internalHashVector(doc, bitsInSimHash)).getArray();
	}
	
	static byte[] hashVector(List<String> terms, int bitsInSimHash) {
		return new SimHash().hash(internalHashVector(terms, bitsInSimHash)).getArray();
	}
	
	public static HashVector internalHashVector(CollectionDocument doc, int bitsInSimHash) {
		List<String> terms = NGramms.tokenize(doc.getFullyCanonicalizedContent());
		return internalHashVector(terms, bitsInSimHash);
	}
	
	public static HashVector internalHashVector(List<String> terms, int bitsInSimHash) {
		return HashVectorSha3.toVector(terms, bitsInSimHash);
	}
	
	private static Set<Integer> docToElementSet(CollectionDocument doc) {
		return NGramms.build8Gramms(doc.getFullyCanonicalizedContent()).stream()
				.map(i -> i.getMd5Hash().hashCode())
				.collect(Collectors.toSet());
	}
}
