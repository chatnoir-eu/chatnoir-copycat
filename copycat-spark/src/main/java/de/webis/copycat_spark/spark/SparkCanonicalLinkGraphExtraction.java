package de.webis.copycat_spark.spark;

import java.io.IOException;
import java.io.Serializable;
import java.net.URL;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.apache.hadoop.io.LongWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.InputFormat;
import org.apache.hadoop.mapred.InputSplit;
import org.apache.hadoop.mapred.JobConf;
import org.apache.hadoop.mapred.RecordReader;
import org.apache.hadoop.mapred.Reporter;
import org.apache.hadoop.mapred.SequenceFileInputFormat;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaHadoopRDD;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.json.JSONException;
import org.json.JSONObject;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.select.Elements;

import com.fasterxml.jackson.databind.ObjectMapper;

import de.webis.chatnoir2.mapfile_generator.inputformats.ClueWeb09InputFormat;
import de.webis.chatnoir2.mapfile_generator.warc.WarcRecord;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import io.anserini.index.transform.JsoupStringTransform;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.SneakyThrows;

public class SparkCanonicalLinkGraphExtraction {

//	@SuppressWarnings("unchecked")
//	public static void main(String[] args) {
////		String corpus = "cc-2017-04";
////		String sampleSize = "0.1";
////		String path = "/corpora/corpus-commoncrawl/CC-MAIN-2017-04-mapfile/data-r-*/data";
//		String corpus = "cc-2015-11";
//		String sampleSize = "0.1";
//		String path = "/corpora/corpus-commoncrawl/CC-MAIN-2015-11-mapfile/data-r-*/data";
//		
//		try (JavaSparkContext context = context()) {
//			Set<String> urlsToKeep = canonicalLinksToKeep(context.textFile("cikm2020/canonical-link-graph/" + corpus + "-canonical-urls-to-count-sample-" + sampleSize));
//			JavaHadoopRDD<Text, Text> rdd = (JavaHadoopRDD<Text, Text>) context.hadoopFile(path, SequenceFileInputFormat.class, Text.class, Text.class);	
//		
//			canonicalLinkedges(rdd, urlsToKeep).saveAsTextFile("cikm2020/canonical-link-graph/" + corpus + "-sample-" + sampleSize);
////			canonicalLinks(rdd).saveAsTextFile("cikm2020/canonical-link-graph/cc-2017-04-canonical-urls");
//		}
//	}

	public static void main(String[] args) {
		try (JavaSparkContext context = context()) {
			JavaRDD<String> tmp = context.textFile("s3a://corpus-copycat/document-representations/cw09");
			JavaPairRDD<LongWritable, WarcRecord> records = context.newAPIHadoopFile("", ClueWeb09InputFormat.class, LongWritable.class, WarcRecord.class, context.hadoopConfiguration());
			
			tmp.sample(false, 0.01).saveAsTextFile("tmp-maik-delete-me.jsonl");
		}
	}
	
	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/canonical-link-graph");
	
		return new JavaSparkContext(conf);
	}
	
	public static JavaRDD<String> canonicalLinks(JavaPairRDD<Text, Text> input) {
		return input.map(i -> toCanonicalUrl(i._1().toString(), i._2().toString()))
				.filter(i -> i != null);
	}
	
	public static JavaRDD<String> canonicalLinkedges(JavaPairRDD<Text, Text> input, Set<String> urlsToKeep) {
		return input.map(i -> toVertex(i._1().toString(), i._2().toString(), urlsToKeep))
				.filter(i -> i != null)
				.map(i -> i.toString());
	}

	@SneakyThrows
	private static String toCanonicalUrl(String internalId, String json) {
		CanonicalLinkGraphEdge ret = toVertex(internalId, json, null);
		
		if(ret == null || ret.getCanonicalLink() == null) {
			return null;
		}
		
		return ret.getCanonicalLink().toString();
	}
	
	@SneakyThrows
	private static CanonicalLinkGraphEdge toVertex(String internalId, String json, Set<String> urlsToKeep) {
		// ignore large files
        if (json.getBytes().length > 1024 * 1024) {
            return null;
        }
        
        
        JSONObject inputJson  = new JSONObject(json);
        final JSONObject metadata = inputJson.getJSONObject("metadata");
        if (null == metadata) {
            throw new JSONException("Missing 'metadata'");
        }

        final JSONObject payload = inputJson.getJSONObject("payload");
        if (null == payload) {
            throw new JSONException("Missing 'payload'");
        }

        final String contentBody = payload.getString("body");
        final String contentEncoding    = payload.getString("encoding");

        if (null == contentEncoding || null == contentBody) {
            throw new JSONException("Missing one of 'payload/[encoding|body]'");
        }

        if (!contentEncoding.equals("plain")) {
            return null;
        }
        
        if(!"response".equals(metadata.getString("WARC-Type"))) {
        	return null;
        }

        String targetUri = metadata.getString("WARC-Target-URI");
        URL canonicalLink = extractCanonicalLinkOrNull(targetUri, contentBody);
        if(canonicalLink == null || (urlsToKeep != null && !urlsToKeep.contains(canonicalLink.toString()))) {
        	return null;
        }
        
        
        String id = internalId;
        if(metadata.has("WARC-TREC-ID")) {
        	id = metadata.getString("WARC-TREC-ID");
        }
        String timestamp = metadata.getString("WARC-Date");
        
        CollectionDocument ret = CollectionDocument.collectionDocument(new JsoupStringTransform().apply(contentBody), id);
        ret.setUrl(new URL(targetUri));
        
		return new CanonicalLinkGraphEdge(ret, canonicalLink, timestamp);
	}
	
	@SneakyThrows
	public static URL extractCanonicalLinkOrNull(String resolveFrom, String contentBody) {
		try {
			return extractCanonicalLinkOrNull(resolveFrom, Jsoup.parse(contentBody));
		} catch(Exception e) {
			return null;
		}
	}
	
	@SneakyThrows
	public static URL extractCanonicalLinkOrNull(String resolveFrom, Document doc) {
		try {
			Elements canonicals = doc.head().select("link[rel=\"canonical\"][href]");
			if(canonicals.size() == 0) {
				return null;
			}
	
			return new URL(new URL(resolveFrom), canonicals.get(0).attr("href"));
		} catch(Exception e) {
			return null;
		}
	}

	@Data
	@NoArgsConstructor
	@AllArgsConstructor
	@SuppressWarnings("serial")
	public static class CanonicalLinkGraphEdge implements Serializable {
		private CollectionDocument doc;
		private URL canonicalLink;
		private String crawlingTimestamp;
		
		@SneakyThrows
		public String toString() {
			return new ObjectMapper().writeValueAsString(this);
		}
		
		@SneakyThrows
		public static CanonicalLinkGraphEdge fromString(String src) {
			return new ObjectMapper().readValue(src, CanonicalLinkGraphEdge.class);
		}
	}

	public static Set<String> canonicalLinksToKeep(JavaRDD<String> input) {
		List<String> ret = input.map(i -> urlFromUrlToCount(i))
				.filter(i -> i != null)
				.distinct().collect();
		
		return new HashSet<>(ret);
	}
	
	@SneakyThrows
	private static final String urlFromUrlToCount(String src) {
		return (String) new ObjectMapper().readValue(src, Map.class).get("url");
	}
}
