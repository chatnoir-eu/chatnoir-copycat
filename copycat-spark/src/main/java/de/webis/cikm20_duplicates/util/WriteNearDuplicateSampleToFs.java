package de.webis.cikm20_duplicates.util;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

import de.webis.cikm20_duplicates.app.SampleNearDuplicates.NdSample;
import de.webis.cikm20_duplicates.spark.SparkEnrichRelevanceTransferPairs;
import de.webis.cikm20_duplicates.spark.SparkCalculateCanonicalLinkGraphEdgeLabels.CanonicalLinkGraphEdge2;
import de.webis.cikm20_duplicates.util.CollectionDocumentUtil.EsDocumentResolver;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import lombok.SneakyThrows;

public class WriteNearDuplicateSampleToFs {
	private static final String
	DIR = "src/main/resources/near-duplicate-sample/cc-2017-04/";
	
	//CREATED BY: de.webis.cikm20_duplicates.app.SampleNearDuplicatesIntegration
	private static final Path INPUT_PATH = Paths.get("src/main/resources/identified-near-duplicates-human-judgments/cw09.jsonl");
	
	
//	FIXME: Kann ich mittels S3 die kompletten near-duplicates berechnen, indem ich die Kandidaten final abteste?
//			Z.B. indem ich auf basis der near-duplicates geeignete blöcke bilde, die dann mit S3-Index getestet werden können?
//			Irgendwie in der Art: Die near-duplicate-Pairs geben mir irgendwie gruppen vor, mit denen ich nochmal nachberechnen kann.
//			Vielleicht über irgendwas transitives?
//			Mit neuen Berechnungen?
//			- für k=0 ist das machbar-> partitioniere Dokumente in den jeweiligen Block
//			- für k=1 muss man sich alle gruppen speichern, in die es verglichen werden muss?
//			  - Kann ich vielleicht die Idee anwenden, die in dem anderen Gedanken (bilde immer ab auf minimum) angewandt wurden?
//					  
//	- Nutze Elasticsearch fuer wahlfreien Zugriff auf Dokumente zur Pruefung
	public static void main(String[] args) {
		int count = 0;
		int wrong = 0;
		for(NdSample sample: samples(INPUT_PATH)) {
			CollectionDocument a = new EsDocumentResolver().loadCollectionDocument(sample.getFirstId());
			CollectionDocument b = new EsDocumentResolver().loadCollectionDocument(sample.getSecondId());
			double tmpScore = -1;
			if(a != null && b != null) {
				tmpScore = SparkEnrichRelevanceTransferPairs.s3Score(a,b);
			}
			
			System.out.println("hemming: " + sample.getHemmingDistance() +"; s3Score: " + sample.getS3Score() +"; mainContentS3Score: " + tmpScore);
			
			if(sample.getS3Score() < 0.82) {
//				System.out.println(sample.getS3Score() + " --> " + sample.getHemmingDistance());
			}
			
			int minHemming = 0;
			if(minHemming >= sample.getHemmingDistance()) {
				count++;
			}
			
			if(sample.getS3Score() < 0.82) {
//				System.out.println(sample.getFirstURL() + " vs " + sample.getSecondURL());
//				System.out.println(sample.getS3Score() + " --> " + sample.getHemmingDistance());
//				System.out.println();
				
//				System.out.println(sample);
				if(minHemming >= sample.getHemmingDistance()) {
					wrong++;
				}
			}
//			System.out.println(sample.getS3Score());
//			persistExample(sample)
		}
		
		System.out.println("---> " + (((double) wrong)/(double) count));
	}
	
	@SneakyThrows
	private static void persistExample(NdSample sample) {
		CollectionDocument src = sample.getFirstDocument();
		CollectionDocument target = sample.getSecondDocument();
		
		Path dir = Paths.get(DIR + src.getId() + "__to__" + target.getId());
		dir.toFile().mkdirs();
		
		Files.write(dir.resolve("metadata"), metadata(sample).getBytes(StandardCharsets.UTF_8));
		Files.write(dir.resolve("src-doc"), src.getFullyCanonicalizedContent().getBytes(StandardCharsets.UTF_8));
		Files.write(dir.resolve("target-doc"), target.getFullyCanonicalizedContent().getBytes(StandardCharsets.UTF_8));
	}
	
	private static String metadata(NdSample sample) {
		return "First: "+ sample.getFirstURL() +"\n" + 
				"Second: "+ sample.getSecondId() +"\n" +
				"S3: " + sample.getS3Score() +"\n";
	}
	
	@SneakyThrows
	private static List<NdSample> samples(Path path) {
		List<String> lines = Files.readAllLines(path);
		List<NdSample> ret = new ArrayList<>();
		
		for(String line: lines) {
			ret.add(NdSample.fromString(line));
		}
		
		return ret;
	}
}
