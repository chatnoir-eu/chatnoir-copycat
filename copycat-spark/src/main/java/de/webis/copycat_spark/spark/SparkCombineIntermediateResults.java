package de.webis.copycat_spark.spark;

import java.util.Map;

import org.apache.commons.lang3.StringUtils;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

import com.fasterxml.jackson.databind.ObjectMapper;

import lombok.SneakyThrows;

public class SparkCombineIntermediateResults {
	
	public static void main(String[] args) {
		try (JavaSparkContext context = context()) {
			JavaRDD<String> input = context.textFile("cikm2020/deduplication/near-duplicates/cw09-cw12");
			
			combineIntermediateResults(input, 5000)
				.saveAsTextFile("cikm2020/results/test-01");
		}
	}

	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/candidates");

		return new JavaSparkContext(conf);
	}
	
	public static JavaRDD<String> combineIntermediateResults(JavaRDD<String> inputRDD, int partitions) {
		return inputRDD
				.map(i -> toCsv(i))
				.filter(i -> i != null)
				.distinct()
				.sortBy(i -> i, Boolean.TRUE, partitions);
	}

	@SneakyThrows
	@SuppressWarnings("unchecked")
	private static String toCsv(String json) {
		if(StringUtils.isBlank(json)) {
			return null;
		}
		
		Map<String, Object> ret = new ObjectMapper().readValue(json, Map.class);
		String firstId = (String) ret.get("firstId");
		String secondId = (String) ret.get("secondId");
		Integer hemmingDistance = (Integer) ret.get("hemmingDistance");
		
		if(firstId == null|| secondId == null || hemmingDistance == null) {
			throw new IllegalArgumentException("Problem with input: '" + json + "'.");
		}
		
		return firstId + "," + secondId + "," + hemmingDistance;
	}
}
