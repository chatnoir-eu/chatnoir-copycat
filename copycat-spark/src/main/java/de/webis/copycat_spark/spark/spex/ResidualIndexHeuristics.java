package de.webis.copycat_spark.spark.spex;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.apache.commons.lang3.tuple.Pair;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.Optional;
import org.codehaus.jackson.map.ObjectMapper;

import com.google.common.collect.ImmutableList;

import de.webis.copycat_spark.spark.spex.ResidualIndex.ResidualIndexEntry;
import de.webis.trec_ndd.spark.DocumentHash;
import de.webis.trec_ndd.spark.S3ScoreOnWord8GrammIndex.S3Score;
import de.webis.trec_ndd.spark.S3ScoreOnWord8GrammIndex.S3ScoreIntermediateResult;
import de.webis.trec_ndd.util.SymmetricPairUtil;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.SneakyThrows;
import scala.Tuple2;

public class ResidualIndexHeuristics {
	
	public static List<ResidualIndexHeuristic> sortedHeuristics(JavaPairRDD<String, DocumentHash> metadata, JavaPairRDD<String, ResidualIndexEntry> residualIndex, double threshold) {
		JavaRDD<ResidualIndexHeuristic> ret = heuristics(metadata, residualIndex);
		ret = ret.filter(i -> i.getUpperS3Bound() >= threshold);
		ret = ret.sortBy(i -> i.docLength, true, 0);
		
		return ret.collect();
	}
	
	public static boolean canBeAboveThreshold(ResidualIndexHeuristic a, ResidualIndexHeuristic b, double threshold) {
		S3ScoreIntermediateResult ret = new S3ScoreIntermediateResult();
		ret.setCommonNGramms(Math.min(a.residualTokens, b.residualTokens));
		
		ret.setLeftMetadata(pseudoHash(a));
		ret.setRightMetadata(pseudoHash(b));
		
		return new S3Score(ret).getS3Score() >= threshold;
	}
	
	private static DocumentHash pseudoHash(ResidualIndexHeuristic a) {
		DocumentHash ret = new DocumentHash();
		ret.setFullyCanonicalizedWord8GrammSetSize(a.getDocLength());
		
		return ret;
	}
	
	public static JavaRDD<ResidualIndexHeuristic> heuristics(JavaPairRDD<String, DocumentHash> metadata, JavaPairRDD<String, ResidualIndexEntry> residualIndex) {
		JavaPairRDD<String, Integer> docToResidualIndexEntries = residualIndex.mapToPair(i -> new Tuple2<>(i._1(), length(i._2())));
		
		return docToResidualIndexEntries.leftOuterJoin(metadata).groupByKey().map(i -> new ResidualIndexHeuristic(i));
	}
	
	private static int length(ResidualIndexEntry entry) {
		Set<String> ret = new HashSet<>(entry.getResidualTokens());
		
		return ret.size();
	}
	
	@Data
	@NoArgsConstructor
	@AllArgsConstructor
	@SuppressWarnings("serial")
	public static class ResidualIndexHeuristic implements Serializable {
		String documentId;
		int residualTokens;
		int docLength;
		
		public Float getUpperS3Bound() {
			return (2*((float)residualTokens))/((float) docLength);
		}
		
		public int getOtherDocRange(double threshold) {
			int maxOtherLength = (int) Math.ceil(((((double)residualTokens)*2)/threshold)- ((double) docLength) + 1d);
			
			return Math.max(0, maxOtherLength - docLength);
		}

		public ResidualIndexHeuristic(Tuple2<String, Iterable<Tuple2<Integer, Optional<DocumentHash>>>> i) {
			this.documentId = i._1();
			List<Tuple2<Integer, Optional<DocumentHash>>> data = ImmutableList.copyOf(i._2().iterator());
			if(data.size() != 1 || !data.get(0)._2.isPresent()) {
				throw new RuntimeException("I cant handle " + data + " for id '" + documentId + "'.");
			}
			
			this.residualTokens = data.get(0)._1();
			this.docLength = (int) data.get(0)._2.get().getFullyCanonicalizedWord8GrammSetSize();
		}
	}

	public static JavaRDD<String> extractCandidates(JavaSparkContext jsc, List<ResidualIndexHeuristic> input, double threshold) {
		List<Integer> tasks = new ArrayList<>();
		
		for(int i=0; i< input.size(); i++) {
			failIfEntryIsInvalid(input, i);
			tasks.add(i);
		}
		
		return jsc.parallelize(tasks)
				.map(i -> runLoopForTask(i, input, threshold));
	}
	
	@SneakyThrows
	private static String runLoopForTask(Integer position, List<ResidualIndexHeuristic> input, double threshold) {
		List<Pair<String, String>> pairsToCalculate = new ArrayList<>();
		int consideredPairs = 0;
		ResidualIndexHeuristic doc = input.get(position);
		int stopAtLength = doc.docLength + doc.getOtherDocRange(threshold);
		
		for(int i= position+1; i< input.size(); i++) {
			consideredPairs++;
			ResidualIndexHeuristic curr = input.get(i);
			
			if(stopAtLength < curr.docLength) {
				break;
			}
			
			if(canBeAboveThreshold(doc, curr, threshold)) {
				pairsToCalculate.add(SymmetricPairUtil.of(doc.getDocumentId(), curr.getDocumentId()));
			}
		}
		
		Map<String, Object> ret = new LinkedHashMap<>();
		ret.put("position", position);
		ret.put("threshold", threshold);
		ret.put("consideredPairs", consideredPairs);
		ret.put("pairsToCalculate", pairsToCalculate);
		
		return new ObjectMapper().writeValueAsString(ret);
	}

	private static void failIfEntryIsInvalid(List<ResidualIndexHeuristic> input, int pos) {
		if (pos > 0) {
			ResidualIndexHeuristic current = input.get(pos);
			ResidualIndexHeuristic previous = input.get(pos-1);
			
			if (current.docLength < previous.docLength) {
				throw new RuntimeException("Can not handle: " + current + " and " + previous);
			}
		}
	}
}
