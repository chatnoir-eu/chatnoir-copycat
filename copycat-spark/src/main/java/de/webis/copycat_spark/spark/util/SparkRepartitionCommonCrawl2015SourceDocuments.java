package de.webis.copycat_spark.spark.util;

import org.apache.hadoop.io.compress.BZip2Codec;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

public class SparkRepartitionCommonCrawl2015SourceDocuments {
	
	private static final String[] CORPORA = new String[] {
			"cc-2015-11-part-0", "cc-2015-11-part-1", "cc-2015-11-part-2",
			"cc-2015-11-part-3", "cc-2015-11-part-4", "cc-2015-11-part-5",
			"cc-2015-11-part-6", "cc-2015-11-part-7", "cc-2015-11-part-8",
			"cc-2015-11-part-9"
	};
	
	public static void main(String[] args) {
		try (JavaSparkContext context = context()) {
			String corpus = "cc-2015-11";
			
			for(String c: CORPORA) {
				context.textFile("cikm2020/document-fingerprints-final/" + c + "-jsonl")
					.repartition(100)
					.saveAsTextFile("cikm2020/document-fingerprints-final/" + c + "-jsonl-repartioned");
			}
			
			JavaRDD<String> input = context.textFile("cikm2020/document-fingerprints-final/" + corpus +"*-jsonl-repartioned");
			
			input.repartition(10000)
				.saveAsTextFile("cikm2020/document-fingerprints-final/" + corpus +"-jsonl.bzip2", BZip2Codec.class);
		}
	}
	
	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/repartition-data");

		return new JavaSparkContext(conf);
	}
}
