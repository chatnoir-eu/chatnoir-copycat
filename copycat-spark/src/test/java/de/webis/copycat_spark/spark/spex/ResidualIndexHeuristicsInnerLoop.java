package de.webis.copycat_spark.spark.spex;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import org.approvaltests.Approvals;
import org.junit.Assert;
import org.junit.Test;

import de.webis.copycat_spark.spark.SparkIntegrationTestBase;
import de.webis.copycat_spark.spark.spex.ResidualIndexHeuristics.ResidualIndexHeuristic;

public class ResidualIndexHeuristicsInnerLoop extends SparkIntegrationTestBase {
	@Test(expected=RuntimeException.class)
	public void testUnsortedListFails() {
		List<ResidualIndexHeuristic> input = Arrays.asList(
			new ResidualIndexHeuristic("id-1", 1, 1),
			new ResidualIndexHeuristic("id-2", 2, 0)
		);
		
		extractCandidates(input, 0.5);
	}
	
	@Test
	public void approveExtractionOfCandidatesForIdenticalLength() {
		List<ResidualIndexHeuristic> input = Arrays.asList(
			new ResidualIndexHeuristic("id-1", 1, 2),
			new ResidualIndexHeuristic("id-2", 1, 2)
		);
		
		List<String> actual = extractCandidates(input, 0.5);
		Approvals.verifyAsJson(actual);
	}
	
	@Test
	public void approveExtractionOfCandidatesWithDifferentLength() {
		List<ResidualIndexHeuristic> input = Arrays.asList(
			new ResidualIndexHeuristic("id-1", 1, 2),
			new ResidualIndexHeuristic("id-2", 10, 10)
		);
		
		List<String> actual = extractCandidates(input, 0.5);
		Approvals.verifyAsJson(actual);
	}
	
	@Test
	public void testExtractionOfCandidatesWithDifferentLength() {
		List<ResidualIndexHeuristic> input = Arrays.asList(
			new ResidualIndexHeuristic("id-1", 1, 2),
			new ResidualIndexHeuristic("id-2", 10, 10),
			new ResidualIndexHeuristic("id-3", 100, 100),
			new ResidualIndexHeuristic("id-4", 1000, 1000),
			new ResidualIndexHeuristic("id-5", 10000, 10000),
			new ResidualIndexHeuristic("id-6", 100000, 100000)
		);
		
		List<String> actual = extractCandidates(input, 0.5);
		
		Assert.assertEquals(6, actual.size());
		for(String i: actual) {
			Assert.assertTrue(i.contains("\"pairsToCalculate\":[]}"));
		}
	}
	
	@Test
	public void approveExtractionOfCandidatesWithSameLength() {
		List<ResidualIndexHeuristic> input = Arrays.asList(
			new ResidualIndexHeuristic("id-1", 1, 2),
			new ResidualIndexHeuristic("id-2", 10, 10),
			new ResidualIndexHeuristic("id-3", 100, 100),
			new ResidualIndexHeuristic("id-4", 99, 101),
			new ResidualIndexHeuristic("id-5", 98, 102),
			new ResidualIndexHeuristic("id-6", 97, 103)
		);

		List<String> actual = extractCandidates(input, 0.5);
		Approvals.verifyAsJson(actual);
	}
	
	@Test
	public void testExtractionOfCandidatesWithSameLength() {
		List<ResidualIndexHeuristic> input = Arrays.asList(
			new ResidualIndexHeuristic("id-1", 1, 2),
			new ResidualIndexHeuristic("id-2", 10, 10),
			new ResidualIndexHeuristic("id-3", 100, 100),
			new ResidualIndexHeuristic("id-4", 48, 101),
			new ResidualIndexHeuristic("id-5", 48, 102),
			new ResidualIndexHeuristic("id-6", 48, 103)
		);

		List<String> actual = extractCandidates(input, 0.5);
		
		Assert.assertEquals(6, actual.size());
		for(String i: actual) {
			Assert.assertTrue(i.contains("\"pairsToCalculate\":[]}"));
		}
	}
	
	private List<String> extractCandidates(List<ResidualIndexHeuristic> input, double threshold) {
		List<String> ret = new ArrayList<>(ResidualIndexHeuristics.extractCandidates(jsc(), input, threshold).collect());
		Collections.sort(ret);
		return ret;
	}
}
