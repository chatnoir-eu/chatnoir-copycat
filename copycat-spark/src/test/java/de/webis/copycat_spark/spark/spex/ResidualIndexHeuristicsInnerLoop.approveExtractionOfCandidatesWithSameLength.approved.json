[
  "{\"position\":0,\"threshold\":0.5,\"consideredPairs\":1,\"pairsToCalculate\":[]}",
  "{\"position\":1,\"threshold\":0.5,\"consideredPairs\":1,\"pairsToCalculate\":[]}",
  "{\"position\":2,\"threshold\":0.5,\"consideredPairs\":3,\"pairsToCalculate\":[{\"left\":\"id-3\",\"right\":\"id-4\",\"value\":\"id-4\",\"key\":\"id-3\"},{\"left\":\"id-3\",\"right\":\"id-5\",\"value\":\"id-5\",\"key\":\"id-3\"},{\"left\":\"id-3\",\"right\":\"id-6\",\"value\":\"id-6\",\"key\":\"id-3\"}]}",
  "{\"position\":3,\"threshold\":0.5,\"consideredPairs\":2,\"pairsToCalculate\":[{\"left\":\"id-4\",\"right\":\"id-5\",\"value\":\"id-5\",\"key\":\"id-4\"},{\"left\":\"id-4\",\"right\":\"id-6\",\"value\":\"id-6\",\"key\":\"id-4\"}]}",
  "{\"position\":4,\"threshold\":0.5,\"consideredPairs\":1,\"pairsToCalculate\":[{\"left\":\"id-5\",\"right\":\"id-6\",\"value\":\"id-6\",\"key\":\"id-5\"}]}",
  "{\"position\":5,\"threshold\":0.5,\"consideredPairs\":0,\"pairsToCalculate\":[]}"
]